### PR TITLE
[ValueTracking] Fix incorrect inferrence about the signbit of sqrt

### LINKED ISSUE #92217: Bad codegen after commit 3aae916
We have an internal test that recently started producing incorrect output which I bisected back to commit 3aae916ff7fe9d0953aa63b2ba1d0e871f6f76fc.

Consider the following code:
```c++
static void init(unsigned char pred, volatile void *data, unsigned size) {
  unsigned char *bytes = (unsigned char *)data;
  for (unsigned i = 0; i != size; ++i) {
    bytes[i] = pred + i;
  }
}
#define INIT(PRED, VAR) init(PRED, &VAR, sizeof(VAR))

#include <x86intrin.h>

int main(int argc, char *argv[])
{
      __m256 id7702;
      INIT(243, id7702);
    __m256 id7701 = _mm256_sqrt_ps(id7702);
  volatile int id7700 = _mm256_movemask_ps(id7701);
  //printf("id7700:%x\n", id7700);
  return id7700;
}
```
If compiled with optimizations and `-mavx` and the resulting binary is run, the return value changes after 3aae916ff7fe9d0953aa63b2ba1d0e871f6f76fc:
```
$ ~/src/upstream/3aae916ff7fe9d0953aa63b2ba1d0e871f6f76fc-linux/bin/clang -O2 -mavx repro.cpp -o repro.bad.out
$ ~/src/upstream/f658d84e01bcdd49e27dc9ef80e1a6cc5f9417fe-linux/bin/clang -O2 -mavx repro.cpp -o repro.good.out
$ ./repro.good.out 
$ echo $?
7
$ ./repro.bad.out 
$ echo $?
0
```

You can also see the difference on godbolt comparing with 18.1.0:
https://godbolt.org/z/KPbK4jcGc

### PR DESCRIPTION
**Pull request by dtcxzyw**
According to IEEE Std 754-2019, `sqrt` returns nan when the input is negative (except for -0). In this case, we cannot make assumptions about sign bit of the result.
BTW, alive2 seems to treat it as poison :( See https://github.com/AliveToolkit/alive2/issues/1037

Fixes https://github.com/llvm/llvm-project/issues/92217

### CODE REVIEW

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -4940,11 +4940,8 @@ void computeKnownFPClass(const Value *V, const APInt &DemandedElts,
       // subnormal input could produce a negative zero output.
       const Function *F = II->getFunction();
       if (Q.IIQ.hasNoSignedZeros(II) ||
-          (F && KnownSrc.isKnownNeverLogicalNegZero(*F, II->getType()))) {
+          (F && KnownSrc.isKnownNeverLogicalNegZero(*F, II->getType())))
         Known.knownNot(fcNegZero);
-        if (KnownSrc.isKnownNeverNaN())
```
Comment (arsenm) [inline llvm/lib/Analysis/ValueTracking.cpp:4945]: && if the source is known positive

Comment (dtcxzyw) [reply llvm/lib/Analysis/ValueTracking.cpp:4945]: Isn't it covered by the above code?

https://github.com/llvm/llvm-project/blob/b27eb0ae8280675fc8fb249d39f1ccafa3ee2187/llvm/lib/Analysis/ValueTracking.cpp#L4932-L4937

Comment (arsenm) [reply llvm/lib/Analysis/ValueTracking.cpp:4945]: Semi-related, this should be using propagateNaN instead of repeating the snan logic

Comment (arsenm) [reply llvm/lib/Analysis/ValueTracking.cpp:4945]: I think we are missing handling of the nnan flag on the sqrt instruction itself

Comment (arsenm) [reply llvm/lib/Analysis/ValueTracking.cpp:4945]: Or that just comes from the RAII thing

Comment (dtcxzyw) [reply llvm/lib/Analysis/ValueTracking.cpp:4945]: > I think we are missing handling of the nnan flag on the sqrt instruction itself

See https://github.com/llvm/llvm-project/blob/f5c8242042c959cbf476778e1d8a874fc8747801/llvm/include/llvm/Analysis/ValueTracking.h#L510-L527

Comment (dtcxzyw) [reply llvm/lib/Analysis/ValueTracking.cpp:4945]: See the test `sqrt_negative_input_nnan`.

**Review by arsenm — APPROVED** [review]

**Comment by nikic** [comment]
@dtcxzyw LLVM 18 does not accept backports anymore.

**Pull request merged**